### PR TITLE
Contract Factory Support

### PR DESCRIPTION
**Pull request by rodrigo-pino**
Transpile Solidity's `new Contract` expressions to Cairo's `deploy` system calls. 

- [x] Encode constructor arguments as raw felt calldata
- [x] Define a placeholder for class hash
- [x] Fill placeholder during behaviour tests
- [x] Fill placeholder during semantic tests
- [x] Fill placeholder during common warp usage
- [x] Allow using salt option for contract creation (only)
- [x] Behavior tests with new
- [x] Semantic tests with new
- [x] Gather and report unrelated bugs found during development:
       - Bug on `tests/behaviour/contracts/new/product.sol` when calling `getX` from arrays
       - Bug when `getX` returns a struct defined globally. Example in `tests/behaviour/conracts/new/many.sol`

### CODE REVIEW

File: src/transpiler.ts
Comment (GuyMcComb) [inline src/transpiler.ts:65]: This can be moved to passes

File: src/passes/newToDeploy.ts
Comment (GuyMcComb) [inline src/passes/newToDeploy.ts:206]: @DomHenderson How do we feel about this being 160bits?

File: src/passes/implicitConversionToExplicit.ts
Comment (rodrigo-pino) [inline src/passes/implicitConversionToExplicit.ts:208]: This is a left over. Marking it here for when adressing changes

File: src/passes/newToDeploy.ts
Comment (rodrigo-pino) [reply src/passes/newToDeploy.ts:206]: I marked it as 160 bits thought in Cairo it can be bigger than that, it would never be an Uint256 (< 251 bits). Also it has the same size as an address. Perhaps making it an address type instead :thinking:

File: tests/behaviour/contracts/new/many.sol
Comment (GuyMcComb) [inline tests/behaviour/contracts/new/many.sol:74]: Does the encode_to_felt work for calldata?

Comment (rodrigo-pino) [reply tests/behaviour/contracts/new/many.sol:74]: It deals with any type. If not pass the right storage type a conversion is triggered.

Comment (GuyMcComb) [reply tests/behaviour/contracts/new/many.sol:74]: Great!

File: src/passes/newToDeploy.ts
Comment (DomHenderson) [reply src/passes/newToDeploy.ts:206]: Address might make the most sense I think. Not keen on it being 160 since with nethersolc there's no tie between the class hash and 160bits

File: tests/behaviour/expectations/semantic_whitelist.ts
Comment (DomHenderson) [inline tests/behaviour/expectations/semantic_whitelist.ts:1610]: I'm a little uncomfortable with how many failing tests there are here for a PR up for merging (though this may also be because I find it hard to tell from these error messages which are expected failures (WILL NOT SUPPORT) and which are unexpected). Ultimately though, I'll leave it up to you and @GuyMcComb as to how many if any need fixing before this PR is merged

File: src/cairoUtilFuncGen/utils/encodeToFelt.ts
Comment (DomHenderson) [inline src/cairoUtilFuncGen/utils/encodeToFelt.ts:147]: is there a reason for `${0}` rather than just `0` here?

File: src/cairoUtilFuncGen/utils/encodeToFelt.ts
Comment (DomHenderson) [inline src/cairoUtilFuncGen/utils/encodeToFelt.ts:40]: Is it worth adding a test for this before merging?

**Review by DomHenderson — COMMENTED** [review]
Looks good overall. The things I've raised are all fairly minor. The one overall suggestion I have is that this (much like most complex features in warp tbh) would benefit from the dense chunks of code being refactored into named and commented functions

File: tests/behaviour/expectations/semantic_whitelist.ts
Comment (GuyMcComb) [reply tests/behaviour/expectations/semantic_whitelist.ts:1610]: Looks like all the tests that fail because of 'value option' should be changed to will not support? And I assume we have created bug tickets for "Fails because it WARP allows a contract to call itself during creation (And it shoudl not)"

File: src/passes/rejectUnsupportedFeatures.ts
Comment (rodrigo-pino) [reply src/passes/rejectUnsupportedFeatures.ts:174]: Yep, I think we cover them all.
There is no reference type that gets constructed with new other than arrays (& bytes). Neither mappings or structs.
And value types does not get constructed with new, except perhaps a contract type which is an address

File: src/passes/newToDeploy.ts
Comment (GuyMcComb) [reply src/passes/newToDeploy.ts:206]: Could we have this as an Address and then we can confirm in the meeting on Wednesday

File: starknet-testnet/server.py
Comment (rodrigo-pino) [reply starknet-testnet/server.py:96]: Ok, I would prefer leaving it for another PR

File: tests/behaviour/expectations/semantic_whitelist.ts
Comment (rodrigo-pino) [reply tests/behaviour/expectations/semantic_whitelist.ts:1610]: Modified those to `WILL NOT SUPPORT due to reason` to be more clear

File: src/cairoUtilFuncGen/utils/encodeToFelt.ts
Comment (rodrigo-pino) [reply src/cairoUtilFuncGen/utils/encodeToFelt.ts:40]: Yep, it is bugged when it's a global struct definition and it is used across contracts. Already added ticket for those on the Notion board

File: src/passes/references/expectedLocationAnalyser.ts
Comment (rodrigo-pino) [reply src/passes/references/expectedLocationAnalyser.ts:135]: Yes it is

**Review by GuyMcComb — APPROVED** [review]

File: tests/behaviour/contracts/new/product.sol
Comment (rodrigo-pino) [reply tests/behaviour/contracts/new/product.sol:48]: I added the task on the notion board

**Pull request merged**
